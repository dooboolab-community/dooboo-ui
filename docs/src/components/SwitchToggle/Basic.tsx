import {FC, useState} from 'react';
import {SwitchToggle, ThemeProvider, ThemeType, Typography} from 'dooboo-ui';

import {View} from 'react-native';

export const Basic: FC<{themeType: ThemeType}> = ({themeType}) => {
  const [isOn, setIsOn] = useState(false);

  return (
    <ThemeProvider initialThemeType={themeType}>
      <View
        style={{
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        <View style={{flexDirection: 'row', alignItems: 'center'}}>
          <Typography.Body1>{isOn ? 'On' : 'Off'}</Typography.Body1>
          <SwitchToggle isOn={isOn} onPress={() => setIsOn(!isOn)} />
        </View>
        <View style={{flexDirection: 'row', alignItems: 'center'}}>
          <Typography.Body1>{!isOn ? 'On' : 'Off'}</Typography.Body1>
          <SwitchToggle
            style={{transform: [{rotate: '180deg'}]}}
            isOn={!isOn}
            onPress={() => setIsOn(!isOn)}
          />
        </View>
      </View>
    </ThemeProvider>
  );
};
