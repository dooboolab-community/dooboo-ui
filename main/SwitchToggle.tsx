import {Animated, StyleProp, TouchableOpacity, ViewStyle} from 'react-native';
import {DoobooTheme, light} from './theme';
import {ReactElement, useEffect, useState} from 'react';

import {isEmptyObject} from './utils';
import styled from '@emotion/native';
import {withTheme} from '@emotion/react';

interface Styles {
  containerStyle?: ViewStyle;
  onElementContainerStyle?: StyleProp<ViewStyle>;
  offElementContainerStyle?: StyleProp<ViewStyle>;
  circleStyle?: ViewStyle;
  buttonStyle?: StyleProp<ViewStyle>;
  circleColorOff?: string;
  circleColorOn?: string;
  backgroundColorOn?: string;
  backgroundColorOff?: string;
}

interface Props {
  testID?: string;
  isOn: boolean;
  theme?: DoobooTheme;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  duration?: number;
  onElement?: ReactElement;
  offElement?: ReactElement;
  onPress?: () => void;
}

// Typing limitation: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/12202

const AnimatedContainer = styled(Animated.View)`
  flex-direction: row;
  align-items: center;
`;

const defaultContainerStyle: ViewStyle = {
  width: 80,
  height: 40,
  borderRadius: 25,
  padding: 5,
};

const defaultCircleStyle: ViewStyle = {
  width: 32,
  height: 32,
  borderRadius: 16,
};

function Component(props: Props): React.ReactElement {
  const {
    testID,
    theme,
    isOn,
    style,
    styles,
    duration = 300,
    onElement,
    offElement,
    onPress,
  } = props;

  const isValidTheme = theme && !isEmptyObject(theme);

  const {primary, disabled, textContrast, placeholder} = isValidTheme
    ? theme
    : light;

  const {
    backgroundColorOn = primary,
    backgroundColorOff = disabled,
    circleColorOn = textContrast,
    circleColorOff = placeholder,
    containerStyle = defaultContainerStyle,
    circleStyle = defaultCircleStyle,
    buttonStyle,
    onElementContainerStyle,
    offElementContainerStyle,
  } = styles ?? {};

  const paddingLeft: number =
    (containerStyle.padding as number) ||
    (containerStyle.paddingLeft as number) ||
    0;

  const paddingRight: number =
    (containerStyle.padding as number) ||
    (containerStyle.paddingRight as number) ||
    0;

  const circlePosXStart = 0;

  const circlePosXEnd =
    ((containerStyle.width ?? defaultContainerStyle.width) as number) -
    ((circleStyle.width ?? defaultCircleStyle.width) as number) -
    (paddingRight + paddingLeft);

  const [animXValue] = useState(new Animated.Value(isOn ? 1 : 0));

  useEffect(() => {
    const runAnimation = (): void => {
      const option = {
        fromValue: isOn ? 0 : 1,
        toValue: isOn ? 1 : 0,
        duration,
        useNativeDriver: false,
      };

      Animated.timing(animXValue, option).start();
    };

    runAnimation();
  }, [animXValue, isOn, duration]);

  const CircleButton = (
    <Animated.View
      style={[
        circleStyle,
        {
          backgroundColor: animXValue.interpolate({
            inputRange: [0.5, 1],
            outputRange: [
              circleColorOff as string | number,
              circleColorOn as string | number,
            ] as string[] | number[],
          }),
        },
        {
          transform: [
            {
              translateX: animXValue.interpolate({
                inputRange: [0, 1],
                outputRange: [
                  circlePosXStart as string | number,
                  circlePosXEnd as string | number,
                ] as string[] | number[],
              }),
            },
          ],
        },
        buttonStyle,
      ]}
    />
  );

  const OnElement = (
    <Animated.View style={[{opacity: animXValue}, onElementContainerStyle]}>
      {onElement}
    </Animated.View>
  );

  const OffElement = (
    <Animated.View
      style={[
        {
          opacity: animXValue.interpolate({
            inputRange: [0, 1],
            outputRange: [1, 0],
          }),
        },
        offElementContainerStyle,
      ]}>
      {offElement}
    </Animated.View>
  );

  return (
    <TouchableOpacity
      testID={testID}
      accessibilityRole="switch"
      style={style}
      onPress={onPress}
      activeOpacity={0.8}>
      <AnimatedContainer
        style={[
          containerStyle,
          {
            paddingLeft,
            paddingRight,
          },
          {
            backgroundColor: animXValue.interpolate({
              inputRange: [0, 1],
              outputRange: [
                backgroundColorOff as string | number,
                backgroundColorOn as string | number,
              ] as string[] | number[],
            }),
          },
        ]}>
        {isOn ? OnElement : OffElement}
        {CircleButton}
      </AnimatedContainer>
    </TouchableOpacity>
  );
}

export const SwitchToggle = withTheme(Component);
