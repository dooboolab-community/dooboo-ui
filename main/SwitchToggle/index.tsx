import {Animated, StyleProp, TouchableOpacity, ViewStyle} from 'react-native';
import React, {ReactElement, useEffect, useState} from 'react';

import styled from '@emotion/native';
import {useTheme} from '@dooboo-ui/theme';

interface Styles {
  container?: ViewStyle;
  onElementContainer?: StyleProp<ViewStyle>;
  offElementContainer?: StyleProp<ViewStyle>;
  circle?: ViewStyle;
  button?: StyleProp<ViewStyle>;
  circleColorOff?: string;
  circleColorOn?: string;
  backgroundColorOn?: string;
  backgroundColorOff?: string;
}

interface Props {
  testID?: string;
  isOn: boolean;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  duration?: number;
  onElement?: any;
  offElement?: any;
  onPress?: () => void;
}

// Typing limitation: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/12202

const AnimatedContainer = styled(Animated.View)`
  flex-direction: row;
  align-items: center;
`;

const defaultContainer: ViewStyle = {
  width: 80,
  height: 40,
  borderRadius: 25,
  padding: 5,
};

const defaultCircle: ViewStyle = {
  width: 32,
  height: 32,
  borderRadius: 16,
};

export function SwitchToggle(props: Props): ReactElement {
  const {
    testID,
    isOn,
    style,
    styles,
    duration = 300,
    onElement,
    offElement,
    onPress,
  } = props;

  const {
    theme: {primary, disabled, textContrast, placeholder},
  } = useTheme();

  const {
    backgroundColorOn = primary,
    backgroundColorOff = disabled,
    circleColorOn = textContrast,
    circleColorOff = placeholder,
    container = defaultContainer,
    circle = defaultCircle,
    button,
    onElementContainer,
    offElementContainer,
  } = styles ?? {};

  const paddingLeft: number =
    (container.padding as number) || (container.paddingLeft as number) || 0;

  const paddingRight: number =
    (container.padding as number) || (container.paddingRight as number) || 0;

  const circlePosXStart = 0;

  const circlePosXEnd =
    ((container.width ?? defaultContainer.width) as number) -
    ((circle.width ?? defaultCircle.width) as number) -
    (paddingRight + paddingLeft);

  const [animXValue] = useState(new Animated.Value(isOn ? 1 : 0));

  useEffect(() => {
    const runAnimation = (): void => {
      const option = {
        fromValue: isOn ? 0 : 1,
        toValue: isOn ? 1 : 0,
        duration,
        useNativeDriver: false,
      };

      Animated.timing(animXValue, option).start();
    };

    runAnimation();
  }, [animXValue, isOn, duration]);

  const CircleButton = (
    <Animated.View
      style={[
        circle,
        {
          backgroundColor: animXValue.interpolate({
            inputRange: [0.5, 1],
            outputRange: [
              circleColorOff as string | number,
              circleColorOn as string | number,
            ] as string[] | number[],
          }),
        },
        {
          transform: [
            {
              translateX: animXValue.interpolate({
                inputRange: [0, 1],
                outputRange: [
                  circlePosXStart as string | number,
                  circlePosXEnd as string | number,
                ] as string[] | number[],
              }),
            },
          ],
        },
        button,
      ]}
    />
  );

  const OnElement = (
    <Animated.View style={[{opacity: animXValue}, onElementContainer]}>
      {onElement}
    </Animated.View>
  );

  const OffElement = (
    <Animated.View
      style={[
        {
          opacity: animXValue.interpolate({
            inputRange: [0, 1],
            outputRange: [1, 0],
          }),
        },
        offElementContainer,
      ]}
    >
      {offElement}
    </Animated.View>
  );

  return (
    <TouchableOpacity
      testID={testID}
      accessibilityRole="switch"
      style={style}
      onPress={onPress}
      activeOpacity={0.8}
    >
      <AnimatedContainer
        style={[
          container,
          {
            paddingLeft,
            paddingRight,
          },
          {
            backgroundColor: animXValue.interpolate({
              inputRange: [0, 1],
              outputRange: [
                backgroundColorOff as string | number,
                backgroundColorOn as string | number,
              ] as string[] | number[],
            }),
          },
        ]}
      >
        {isOn ? OnElement : OffElement}
        {CircleButton}
      </AnimatedContainer>
    </TouchableOpacity>
  );
}
