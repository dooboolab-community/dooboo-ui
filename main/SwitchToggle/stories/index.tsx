import type {ReactElement} from 'react';
import React from 'react';

import SwitchToggleDefaultStory from './DefaultStory';
import {ThemeProvider} from '@dooboo-ui/theme';
import {storiesOf} from '@storybook/react-native';

/**
 * Below are stories for web
 */
export default {
  title: 'SwitchToggle',
};

export const toStorybook = (): ReactElement => <SwitchToggleDefaultStory />;

toStorybook.story = {
  name: 'switch toggle',
};

/**
 * Below are stories for app
 */
storiesOf('SwitchToggle', module)
  .add('switch toggle - light', () => (
    <ThemeProvider initialThemeType="light">
      <SwitchToggleDefaultStory />
    </ThemeProvider>
  ))
  .add('switch toggle - dark', () => (
    <ThemeProvider initialThemeType="dark">
      <SwitchToggleDefaultStory />
    </ThemeProvider>
  ));
