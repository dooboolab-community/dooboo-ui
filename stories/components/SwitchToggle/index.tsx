import StatusBarBrightness from '../../../main/components/StatusbarBrightness';
import SwitchToggleDefaultStory from './DefaultStory';
import {ThemeProvider} from '@dooboo-ui/theme';
import {storiesOf} from '@storybook/react-native';

storiesOf('SwitchToggle', module)
  .add('Basic - light', () => (
    <ThemeProvider initialThemeType="light">
      <>
        <StatusBarBrightness />
        <SwitchToggleDefaultStory />
      </>
    </ThemeProvider>
  ))
  .add('Basic - dark', () => (
    <ThemeProvider initialThemeType="dark">
      <SwitchToggleDefaultStory />
    </ThemeProvider>
  ));
