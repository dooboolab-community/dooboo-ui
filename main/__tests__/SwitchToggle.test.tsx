import React, {ReactElement} from 'react';

import {SwitchToggle} from '../SwitchToggle';
import {TouchableOpacity} from 'react-native';
import renderer from 'react-test-renderer';

const createTestProps = (
  obj?: Record<string, unknown>,
): Record<string, unknown> => ({
  ...obj,
});

let props;
let component: ReactElement;

describe('[SwitchToggle]', (): void => {
  props = createTestProps({
    testID: 'SWITCH_ID',
    onPress: jest.fn(),
  });

  component = <SwitchToggle {...props} />;

  describe('Rendering', () => {
    it('should render without crashing', () => {
      const rendered = renderer.create(component);

      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });

    it('should render type === 0', () => {
      props = {
        ...props,
        type: 0,
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);

      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });

    it('should render `containerStyle` and `containerStyle.padding`', () => {
      props = {
        ...props,
        type: 1,
        containerStyle: {
          padding: 40,
        },
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);

      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });

    it('should render `containerStyle` and `circleWidth`', () => {
      props = {
        ...props,
        type: 1,
        containerStyle: {
          padding: 40,
          width: 40,
        },
        circleStyle: {
          width: 20,
        },
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);

      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });

    it('should render `containerStyle` and `circleWidth` without padding', () => {
      props = {
        ...props,
        type: 1,
        containerStyle: {
          padding: 0,
          width: 40,
        },
        circleStyle: {
          width: 20,
        },
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);

      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });

    it('should render when `containerStyle` is not defined', () => {
      props = {
        ...props,
        type: 1,
        containerStyle: undefined,
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);

      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });

    it('should render RTL', () => {
      props = {
        ...props,
        type: 1,
        RTL: true,
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);

      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });
  });

  describe('Accessibility', (): void => {
    it('should render AccessibilityLabel', () => {
      props = {
        ...props,
        type: 1,
        accessibilityLabel: 'Label for test accessibility',
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);

      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });

    it('should render AccessibilityRole', () => {
      props = {
        ...props,
        type: 1,
        accessibilityRole: 'switch',
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);

      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });

    it('should render AccessibilityHint', () => {
      props = {
        ...props,
        type: 1,
        accessibilityHint: 'Accessibility hint can be a string or empty',
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);

      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });

    it('should render AccessibilityState', () => {
      props = {
        ...props,
        type: 1,
        accessibilityState: {
          disabled: false,
          selected: true,
          checked: false,
          busy: false,
          expanded: false,
        },
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);

      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });

    it('should render AccessibilityValue', () => {
      props = {
        ...props,
        type: 1,
        accessibilityValue: {
          min: 0,
          max: 1,
          now: 0,
          text: 'optional text',
        },
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);

      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });
  });

  it('should render custom `colors` and `duration`', () => {
    props = {
      ...props,
      backgroundColorOn: 'red',
      backgroundColorOff: 'red',
      circleColorOn: 'red',
      circleColorOff: 'red',
      duration: 500,
    };

    component = <SwitchToggle {...props} />;

    const rendered = renderer.create(component);

    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  describe('Interaction', (): void => {
    it('should simulate onPress', () => {
      const rendered = renderer.create(component);
      const switchToggle = rendered.root.findByType(TouchableOpacity);

      switchToggle.props.onPress();

      expect(props.onPress).toHaveBeenCalled();
    });

    it('should toggle switch on/off when pressed', () => {
      // eslint-disable-next-line @typescript-eslint/no-shadow
      const props = {
        switchOn: false,
        onPress: () => (props.switchOn = !props.switchOn),
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);
      const switchToggle = rendered.root.findByType(TouchableOpacity);

      expect(props.switchOn).toBeFalsy();

      switchToggle.props.onPress();

      expect(props.switchOn).toBeTruthy();

      switchToggle.props.onPress();

      expect(props.switchOn).toBeFalsy();
    });

    it('should toggle switchOn to `false` on pressed', () => {
      // eslint-disable-next-line @typescript-eslint/no-shadow
      const props = {
        switchOn: true,
        onPress: jest.fn(),
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);
      const switchToggle = rendered.root.findByType(TouchableOpacity);

      switchToggle.props.onPress();

      expect(props.switchOn).toBeTruthy();

      switchToggle.props.onPress();

      expect(props.onPress).toHaveBeenCalled();
    });
  });
});
