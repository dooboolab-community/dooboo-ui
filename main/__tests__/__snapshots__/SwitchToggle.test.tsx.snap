// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SwitchToggle] Accessibility should render AccessibilityHint 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(196, 196, 196, 1)",
        "borderRadius": 25,
        "flexDirection": "row",
        "height": 40,
        "marginTop": 16,
        "padding": 5,
        "width": 80,
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(-37, -37, -37, 1)",
          "transform": Array [
            Object {
              "translateX": 10,
            },
          ],
          "width": 20,
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;

exports[`[SwitchToggle] Accessibility should render AccessibilityLabel 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(196, 196, 196, 1)",
        "borderRadius": 25,
        "flexDirection": "row",
        "height": 40,
        "marginTop": 16,
        "padding": 5,
        "width": 80,
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(-37, -37, -37, 1)",
          "transform": Array [
            Object {
              "translateX": 10,
            },
          ],
          "width": 20,
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;

exports[`[SwitchToggle] Accessibility should render AccessibilityRole 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(196, 196, 196, 1)",
        "borderRadius": 25,
        "flexDirection": "row",
        "height": 40,
        "marginTop": 16,
        "padding": 5,
        "width": 80,
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(-37, -37, -37, 1)",
          "transform": Array [
            Object {
              "translateX": 10,
            },
          ],
          "width": 20,
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;

exports[`[SwitchToggle] Accessibility should render AccessibilityState 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(196, 196, 196, 1)",
        "borderRadius": 25,
        "flexDirection": "row",
        "height": 40,
        "marginTop": 16,
        "padding": 5,
        "width": 80,
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(-37, -37, -37, 1)",
          "transform": Array [
            Object {
              "translateX": 10,
            },
          ],
          "width": 20,
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;

exports[`[SwitchToggle] Accessibility should render AccessibilityValue 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(196, 196, 196, 1)",
        "borderRadius": 25,
        "flexDirection": "row",
        "height": 40,
        "marginTop": 16,
        "padding": 5,
        "width": 80,
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(-37, -37, -37, 1)",
          "transform": Array [
            Object {
              "translateX": 10,
            },
          ],
          "width": 20,
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;

exports[`[SwitchToggle] Rendering should render \`containerStyle\` and \`circleWidth\` 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(196, 196, 196, 1)",
        "flexDirection": "row",
        "padding": 40,
        "width": 40,
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(-37, -37, -37, 1)",
          "transform": Array [
            Object {
              "translateX": 80,
            },
          ],
          "width": 20,
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;

exports[`[SwitchToggle] Rendering should render \`containerStyle\` and \`circleWidth\` without padding 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(196, 196, 196, 1)",
        "flexDirection": "row",
        "padding": 0,
        "width": 40,
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(-37, -37, -37, 1)",
          "transform": Array [
            Object {
              "translateX": "NaN[object Object]",
            },
          ],
          "width": 20,
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;

exports[`[SwitchToggle] Rendering should render \`containerStyle\` and \`containerStyle.padding\` 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(196, 196, 196, 1)",
        "flexDirection": "row",
        "padding": 40,
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(-37, -37, -37, 1)",
          "borderRadius": 16,
          "height": 32,
          "transform": Array [
            Object {
              "translateX": NaN,
            },
          ],
          "width": 32,
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;

exports[`[SwitchToggle] Rendering should render RTL 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(196, 196, 196, 1)",
        "borderRadius": 25,
        "flexDirection": "row",
        "height": 40,
        "marginTop": 16,
        "padding": 5,
        "width": 80,
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(-37, -37, -37, 1)",
          "transform": Array [
            Object {
              "translateX": 10,
            },
          ],
          "width": 20,
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;

exports[`[SwitchToggle] Rendering should render type === 0 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(196, 196, 196, 1)",
        "borderRadius": 25,
        "flexDirection": "row",
        "height": 40,
        "marginTop": 16,
        "padding": 5,
        "width": 80,
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(-37, -37, -37, 1)",
          "borderRadius": 16,
          "height": 32,
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
          "width": 32,
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;

exports[`[SwitchToggle] Rendering should render when \`containerStyle\` is not defined 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(196, 196, 196, 1)",
        "borderRadius": 25,
        "flexDirection": "row",
        "height": 40,
        "marginTop": 16,
        "padding": 5,
        "width": 80,
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(-37, -37, -37, 1)",
          "transform": Array [
            Object {
              "translateX": 10,
            },
          ],
          "width": 20,
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;

exports[`[SwitchToggle] Rendering should render without crashing 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(196, 196, 196, 1)",
        "borderRadius": 25,
        "flexDirection": "row",
        "height": 40,
        "marginTop": 16,
        "padding": 5,
        "width": 80,
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(-37, -37, -37, 1)",
          "borderRadius": 16,
          "height": 32,
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
          "width": 32,
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;

exports[`[SwitchToggle] should render custom \`colors\` and \`duration\` 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(255, 0, 0, 1)",
        "borderRadius": 25,
        "flexDirection": "row",
        "height": 40,
        "marginTop": 16,
        "padding": 5,
        "width": 80,
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(255, 0, 0, 1)",
          "transform": Array [
            Object {
              "translateX": 10,
            },
          ],
          "width": 20,
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;
